powerlevel9k.zsh-theme

```diff
@@ -610,6 +609,28 @@ prompt_dir() {
           current_path=$(truncatePathFromRight "$(pwd | sed -e "s,^$HOME,~,")" )
         fi
       ;;
+      truncate_with_folder_marker)
+        local last_marked_folder marked_folder zero
+        set_default POWERLEVEL9K_SHORTEN_FOLDER_MARKER ".shorten_folder_marker"
+
+        for marked_folder in $(upsearch $POWERLEVEL9K_SHORTEN_FOLDER_MARKER); do
+          if [[ "$marked_folder" == "/" ]]; then
+            # If we reached root folder, stop upsearch.
+            current_path="/"
+          elif [[ "$marked_folder" == "$HOME" ]]; then
+            # If we reached home folder, stop upsearch.
+            current_path="~"
+          elif [[ "${marked_folder%/*}" == $last_marked_folder ]]; then
+            current_path="${current_path%/}/${marked_folder##*/}"
+          else
+            current_path="${current_path%/}/$POWERLEVEL9K_SHORTEN_DELIMITER/${marked_folder##*/}"
+          fi
+          last_marked_folder=$marked_folder
+        done
+
+        zero='%([BSUbfksu]|([FB]|){*})'
+        current_path=$current_path${PWD:${#${(S%%)last_marked_folder//$~zero/}}}
+      ;;
```

Comment:
@bhilburn Thanks for pointing me to this again. This is a quite complicated pattern to count the visible characters in the prompt. In our case we count the characters in `$last_marked_folder` (which should not contain any non-visible characters) and replace the amount of characters in `$PWD` from the left with nothing. IMHO this could be done way easier (just replace the characters from the left), but I think @BenoitAverty just took this pattern from the `truncate_with_package_name` strategy. As far as I can see this pattern originates from [barts prompt](https://opensource.apple.com/source/zsh/zsh-55/zsh/Functions/Prompts/prompt_bart_setup). For more information about this pattern, see [this answer](http://stackoverflow.com/a/40855342/5586433) on StackOverflow.

Regarding the performance of `upsearch` function. For me it looks quite solid. Nothing to optimize here.